src/main/java/org/folio/rest/transaction/EncumbranceHandler.java

```diff
@@ -126,19 +140,125 @@ String createPermanentTransactionsQuery() {
         + "ON CONFLICT DO NOTHING;", getFullTransactionTableName(), getTemporaryTransactionTable());
   }
 
-  private Future<Tx<List<Transaction>>> updatePermanentTransactions(Tx<List<Transaction>> tx) {
+  private Future<Tx<List<Transaction>>> processTransactionsUponUpdate(Tx<List<Transaction>> tx) {
+
+    List<Transaction> tempTransactions = tx.getEntity();
+
+    return getPermanentTransactions(getSummaryId(tempTransactions.get(0)))
+      .compose(existingTransactions -> updateBudgetsTotals(tx, existingTransactions))
+      .compose(this::updatePermanentTransactions);
+  }
+
+  private Future<Tx<List<Transaction>>> updateBudgetsTotals(Tx<List<Transaction>> tx, List<Transaction> existingTransactions) {
+    return getBudgets(tx)
+      .map(budgets -> getUpdateBudgetsTotals(existingTransactions, tx.getEntity(), budgets))
+      .compose(budgets -> updateBudgets(tx, budgets));
+  }
+
+  private Future<Tx<List<Transaction>>> updateBudgets(Tx<List<Transaction>> tx, List<Budget> budgets) {
     Promise<Tx<List<Transaction>>> promise = Promise.promise();
+    String sql = "UPDATE " + getFullTableName(getTenantId(), BUDGET_TABLE) + " AS budgets " +
+      "SET jsonb = b.jsonb FROM (VALUES  " + getValues(budgets.stream().map(JsonObject::mapFrom).collect(Collectors.toList())) + ") AS b (id, jsonb) " +
+      "WHERE b.id::uuid = budgets.id;";
+    tx.getPgClient().execute(tx.getConnection(), sql, reply -> {
+      if (reply.failed()) {
+        handleFailure(promise, reply);
+      } else {
+        promise.complete(tx);
+      }
+    });
+    return promise.future();
+  }
+
+  private String getValues(List<JsonObject> entities) {
+    return entities.stream().map(entity -> "('" + entity.getString("id") + "', '" + entity.encode() + "'::json)").collect(Collectors.joining(","));
+  }
+
+  private Future<List<Budget>> getBudgets(Tx<List<Transaction>> tx) {
+    Promise<List<Budget>> promise = Promise.promise();
+    String sql = "SELECT DISTINCT ON (budgets.id) budgets.jsonb " +
```

Comment:
There can be budgets for future FYs too, won't they be returned as well for same fundID?

Reply:
I think there might be retrieved wrong budgets if order was reopened. To avoid that better join  `order_transaction_summaries`.